egs/wsj/s5/utils/data/perturb_speed_to_allowed_lengths.py

```diff
@@ -0,0 +1,298 @@
+                                   speed, one with a higher speed (not more than
+                                   factor% faster) and one with a lower speed
+                                   (not more than factor% slower)""")
+    parser.add_argument('factor', type=float, default=12,
+                        help='Spacing (in percentage) between allowed lengths.')
+    parser.add_argument('srcdir', type=str,
+                        help='path to source data dir')
+    parser.add_argument('dir', type=str, help='output dir')
+    parser.add_argument('--coverage-factor', type=float, default=0.05,
+                        help="""Percentage of durations not covered from each
+                             side of duration histogram.""")
+    parser.add_argument('--frame-shift', type=int, default=10,
+                        help="""Frame shift in milliseconds.""")
+    parser.add_argument('--frame-length', type=int, default=25,
+                        help="""Frame length in milliseconds.""")
+    parser.add_argument('--frame-subsampling-factor', type=int, default=3,
+                        help="""Chain frame subsampling factor.
+                             See steps/nnet3/chain/train.py""")
+    parser.add_argument('--speed-perturb',  type=bool, default=True,
+                        help="""If false, no speed perturbation will occur, i.e.
+                             only 1 copy of each utterance will be
+                             saved, which is modified to have an allowed length
+                             by using extend-wav-with-silence.""")
+    args = parser.parse_args()
+    return args
+
+class Utterance:
+    """ This class represents a Kaldi utterance
+        in a data directory like data/train
+    """
+
+    def __init__(self, uid, wavefile, speaker, transcription, dur):
+        self.wavefile = (wavefile if wavefile.rstrip().endswith('|') else
+                         'cat {} |'.format(wavefile))
+        self.speaker = speaker
+        self.transcription = transcription
+        self.id = uid
+        self.dur = float(dur)
+
+    def to_kaldi_utt_str(self):
+        return self.id + " " + self.transcription
+
+    def to_kaldi_wave_str(self):
+        return self.id + " " + self.wavefile
+
+    def to_kaldi_dur_str(self):
+        return "{} {:0.3f}".format(self.id, self.dur)
+
+
+def read_kaldi_datadir(dir):
+    """ Read a data directory like
+        data/train as a list of utterances
+    """
+
+    utterances = []
+    wav_scp = read_kaldi_mapfile(os.path.join(dir, 'wav.scp'))
+    text = read_kaldi_mapfile(os.path.join(dir, 'text'))
+    utt2dur = read_kaldi_mapfile(os.path.join(dir, 'utt2dur'))
+    utt2spk = read_kaldi_mapfile(os.path.join(dir, 'utt2spk'))
+    for utt in wav_scp:
+        if utt in text and utt in utt2dur and utt in utt2spk:
+            utterances.append(Utterance(utt, wav_scp[utt], utt2spk[utt],
+                                  text[utt], utt2dur[utt]))
+        else:
```

Comment:
Try to add some count like num_success, num_fail and print them. Don't print warning for every utterance. There should also be a special check that the input data directory has been passed through extract_wav_segments script. Otherwise it would generate a lot of warnings without any information on what actually is the issue. 
In additional also check how many utterances you got compared to the ones in wav_scp, and exit if its zero or less than some percentage.